os/pm/pm_metrics.c

```diff
@@ -17,117 +17,276 @@
+ *   domain.
+ *
+ * Input parameters:
+ *   domain_id - the ID of domain registered with PM.
+ *
+ * Returned value:
+ *   None
+ *
+ ****************************************************************************/
+void pm_domain_metricinitialize(int domain_id)
+{
+	if (g_pm_metrics) {
+		g_pm_metrics->dtime[domain_id] = clock_systimer();
+		g_pm_metrics->suspend_ticks[domain_id] = 0;
+	}
 }
 
-void pm_prune_history(sq_queue_t *q)
+/****************************************************************************
+ * Name: pm_suspend_metrics_update
+ *
+ * Description:
+ *   This function is called inside pm_suspend. It note the timestamp (in ticks) of
+ *   suspended domain.
+ *
+ * Input parameters:
+ *   domain_id - the ID of domain registered with PM.
+ *
+ * Returned value:
+ *   None
+ *
+ ****************************************************************************/
+void pm_suspend_metrics_update(int domain_id)
 {
-	sq_queue_t *queue = q;
-	struct pm_statechange_s *sc = NULL;
-	struct timespec ts;
-	time_t cur_time;
-	clock_gettime(CLOCK_REALTIME, &ts);
-	cur_time = ts.tv_sec;
+	if (g_pm_metrics && (g_pmglobals.suspend_count[domain_id] == 0)) {
+		g_pm_metrics->dtime[domain_id] = clock_systimer();
+	}
+}
 
-	sq_entry_t *head = sq_peek(q);
+/****************************************************************************
+ * Name: pm_resume_metrics_update
+ *
+ * Description:
+ *   This function is called inside pm_resume. Before resuming domain, it counts
+ *   amount of time (in ticks) the given domain was suspended.
+ * 
+ * Input parameters:
+ *   domain_id - the ID of domain registered with PM.
+ *
+ * Returned value:
+ *   None
+ *
+ ****************************************************************************/
+void pm_resume_metrics_update(int domain_id)
+{
+	clock_t now;
+	if (g_pm_metrics && (g_pmglobals.suspend_count[domain_id] == 1)) {
+		now = clock_systimer();
+		g_pm_metrics->suspend_ticks[domain_id] += now - g_pm_metrics->dtime[domain_id];
+		g_pm_metrics->dtime[domain_id] = now;
+	}
+}
 
-	for (;;) {
-		/* REmove all but one expired node */
-		if (head && head->flink) {
-			sc = (struct pm_statechange_s *)head->flink;
-			if (sc->timestamp < (cur_time - CONFIG_PM_METRICS_DURATION)) {
-				sq_rem(head, queue);
-				kmm_free(head);
-				head = sq_peek(queue);
-			} else {
-				break;
+/****************************************************************************
+ * Name: pm_idle_metrics_update
+ *
+ * Description:
+ *   This function is called inside pm_idle. It counts the frequency of domain, which
+ *   make board unable to go into sleep during idle cpu time.
+ * 
+ * Input parameters:
+ *   None
+ *
+ * Returned value:
+ *   None
+ *
+ ****************************************************************************/
+void pm_idle_metrics_update(void)
+{
+	int index;
+	if (g_pm_metrics && (g_pmglobals.recommended != PM_SLEEP)) {
+		g_pm_metrics->total_suspend_counts++;
```

Comment:
do we need total_suspend_count?

it seems that it can be calculated as the total measurement time of pm_metrics.

Reply:
total_suspend_count represents no. of times idle thread failed to make board sleep cause of suspended domain

Reply:
let's add report about that top of the report